scraper/tfs/__init__.py

```diff
@@ -0,0 +1,187 @@
+
+        logger.debug('Retrieving Projects for Project Collection: {collection_name}'.format(collection_name=collection.name))
+        # Retrieves all projects in the project collection
+        projects = collection_client.get_projects(top=HARD_CODED_TOP)
+        # get_projects only gets the project references, have to call get_project_history_entries to get last update info for projects
+        # Only calling this once per collection as its an expensive API call, wil refactor later if there is a better API call to use
+        collection_history_list = collection_client.get_project_history_entries()
+        for project in projects:
+
+            # get_projects only gets team project ref objects,
+            # have to call get_project to get the team project object which includes the TFS Web Url for the project
+            logger.debug('Retrieving Team Project for Project: {project_name}'.format(project_name=project.name))
+            projectInfo = collection_client.get_project(project.id, True, True)
+
+            tfsProject = TFSProject(projectInfo, collection)
+
+            logger.debug('Retrieving Last Updated and Created Info for Project: {project_name}'.format(project_name=project.name))
+            tfsProject.projectLastUpdateInfo = get_project_last_update_time(collection_history_list, project.id)
+            tfsProject.projectCreateInfo = get_project_create_time(collection_history_list, project.id)
+            project_list.append(tfsProject)
+
+    return project_list
+
+
+def get_git_repos(url, token, collection, project):
+    """
+    Returns a list of all git repos for the supplied project within the supplied collection
+    """
+    git_client = create_tfs_git_client('{url}/{collection_name}'.format(url=url, collection_name=collection.name), token)
+    logger.debug('Retrieving Git Repos for Project: {project_name}'.format(project_name=project.name))
+    return git_client.get_repositories(project.id)
+
+
+def get_tfvc_repos(url, token, collection, project):
+    """
+    Returns a list of all tfvc branches for the supplied project within the supplied collection
+    """
+    branch_list = []
+    tfvc_client = create_tfs_tfvc_client('{url}/{collection_name}'.format(url=url, collection_name=collection.name), token)
+
+    logger.debug('Retrieving Tfvc Branches for Project: {project_name}'.format(project_name=project.name))
+    branches = tfvc_client.get_branches(project.id, True, True, False, True)
+    if branches:
+        branch_list.extend(branches)
+    else:
+        logger.debug('No Tfvcc Branches in Project: {project_name}'.format(project_name=project.name))
+
+    return branch_list
+
+
+def get_project_last_update_time(collection_history_list, projectId):
+    sorted_history_list = sorted(collection_history_list, key=lambda x: x.last_update_time, reverse=True)
+    project_last_update_info = next((x for x in sorted_history_list if x.id == projectId))
```

Comment:
could collapse this line with the return as `return next((x for x in sorted_history_list if x.id == projectId))` (like `get_project_create_time(...)`

Reply:
Agree. Done.